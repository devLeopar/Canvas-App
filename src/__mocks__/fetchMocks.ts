const mockJsonResponse = (data: any) => {
  return {
    json: () => Promise.resolve(data),
    headers: new Headers(),
    ok: true,
    redirected: false,
    status: 200,
    statusText: "OK",
    type: "default",
    url: "http://mock.url",
    clone: jest.fn(),
    blob: jest.fn(),
    formData: jest.fn(),
    text: jest.fn(),
    arrayBuffer: jest.fn(),
    body: jest.fn(),
    bodyUsed: false,
  };
};

const setupFetchMock = () => {
  global.fetch = jest.fn().mockImplementation((url) => {
    switch (url) {
      case "/static/entities.json":
        return mockJsonResponse([
          { id: 1, name: "test1" },
          { id: 2, name: "test2" },
        ]);
      case "/static/coords.json":
        return mockJsonResponse([
          { id: 1, x: 0, y: 100 },
          { id: 2, x: 50, y: 150 },
        ]);
      default:
        throw new Error(`Unhandled request: ${url}`);
    }
  });
};

export { setupFetchMock };
