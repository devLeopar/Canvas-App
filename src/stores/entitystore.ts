import { observable, action, computed, autorun, makeObservable } from "mobx";
import { EntityObject } from "../types";
import mergeEntitiesWithCoords from "../utils/mergeEntitiesWithCoords";
import { fetchCoords, fetchEntities } from "../services/entityService";

const hasLocalStorage = typeof window !== "undefined" && window.localStorage;

export class Entity {
  id = Math.random();
  name = "Entity";
  x = 0;
  y = 0;

  constructor(json: EntityObject) {
    makeObservable(this, {
      name: observable,
      x: observable,
      y: observable,
      asJson: computed,
    });
    Object.assign(this, json);
  }

  get asJson() {
    const { id, name, x, y } = this;
    return { id, name, x, y };
  }
}

export class EntityStore {
  entities: Entity[] = [];

  constructor() {
    makeObservable(this, {
      entities: observable,
      loadJson: action,
      addEntity: action,
      loadFromLocalStorage: action,
      asJson: computed,
    });
    autorun(this.saveToLocalStorageReaction, { delay: 200 });
  }

  async fetchData() {
    const entities = await fetchEntities();
    const coords = await fetchCoords();

    const mergedData = mergeEntitiesWithCoords(entities, coords);
    this.loadJson(mergedData);
  }

  updatePosition(entityId: number, x: number, y: number) {
    const entity = this.entities.find((e) => e.id === entityId);
    if (entity) {
      entity.x = x;
      entity.y = y;
    }
  }

  loadJson(json: EntityObject[]) {
    this.entities = json.map((entityData) => new Entity(entityData));
  }

  addEntity(name: string, x: number, y: number) {
    this.entities.push(
      new Entity({
        name,
        x,
        y,
      })
    );
  }

  saveToLocalStorageReaction = () => {
    if (hasLocalStorage) {
      window.localStorage.setItem(
        "domain-model-app",
        JSON.stringify(this.asJson)
      );
    }
  };

  loadFromLocalStorage() {
    const data =
      hasLocalStorage && window.localStorage.getItem("domain-model-app");
    if (data) {
      this.loadJson(JSON.parse(data));
    } else {
      this.loadJson([]);
    }
  }

  get asJson() {
    return this.entities.map((e) => e.asJson);
  }
}
