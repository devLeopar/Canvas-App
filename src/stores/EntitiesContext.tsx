import { createContext } from "react";
import { EntityStore } from "./entitystore";

const entityStore = new EntityStore();
export const EntitiesContext = createContext<EntityStore>(entityStore);
