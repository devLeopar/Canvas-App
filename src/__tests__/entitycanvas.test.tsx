import React from "react";
import { render, act } from "@testing-library/react";
import { EntitiesContext } from "../stores/EntitiesContext";
import { EntityStore } from "../stores/entitystore";
import { setupFetchMock } from "../__mocks__/fetchMocks";
import EntityCanvas from "../components/entitycanvas";

describe("<EntityCanvas />", () => {
  let testEntityStore: EntityStore;

  // Setting up fetch mock
  beforeEach(() => {
    testEntityStore = new EntityStore();
    setupFetchMock();
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  // Check if it match the snapshot and make sure its sizes aren't changed
  it("renders correctly", () => {
    const { asFragment } = render(
      <EntitiesContext.Provider value={testEntityStore}>
        <EntityCanvas />
      </EntitiesContext.Provider>
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it("renders the correct number of entities", async () => {
    let draggableEntities;
    await act(async () => {
      const rendered = render(
        <EntitiesContext.Provider value={testEntityStore}>
          <EntityCanvas />
        </EntitiesContext.Provider>
      );
      draggableEntities = await rendered.findAllByTestId("draggable-entity");
    });
    expect(draggableEntities).toHaveLength(testEntityStore.entities.length);
  });

  it("fetches data on mount", async () => {
    await act(async () => {
      render(
        <EntitiesContext.Provider value={testEntityStore}>
          <EntityCanvas />
        </EntitiesContext.Provider>
      );
    });

    // Ensure that the mocked fetch was called
    expect(global.fetch).toHaveBeenCalledTimes(2); // Assuming fetchData makes two fetch calls, one for entities and another for coords
  });
});
