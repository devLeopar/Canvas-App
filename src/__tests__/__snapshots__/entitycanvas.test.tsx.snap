// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<EntityCanvas /> renders correctly 1`] = `
<DocumentFragment>
  <div
    style="position: relative; width: 800px; height: 600px; background-color: rgb(244, 244, 244); border: 1px solid #ddd; border-radius: 4px; margin: 20px 0px;"
  />
</DocumentFragment>
`;
